config/secrets.yml

```diff
@@ -12,14 +12,20 @@
 
 development:
   secret_key_base: e00a3f09f602f74cb64ed7daead6cf63bdaabb9aa6247443bbcb88be36f3e363443c3e1c448809e714dbd8c5626f2c9f8da3cbc244a8ae3ce1f7adddb0a1f88b
+  api_auth_secret: <%= ENV.fetch("API_AUTH_SECRET", "no-api-secret-set") %>
+  enroll_graphql_endpoint: <%= ENV.fetch("ENROLL_GRAPHQL_ENDPOINT", "http://localhost:3001/graphql") %>
 
 test:
   secret_key_base: ac1e649fcb1f5e6ca0afbc18a83a1851ec5e747f43ad29023e9e53120e069e879b7daaf8f4577e775c730a0a7665b0f30c77dd4505ba12bd8cf0432aeb641a05
+  api_auth_secret: <%= ENV.fetch("API_AUTH_SECRET", "no-api-secret-set") %>
+  enroll_graphql_endpoint: <%= ENV.fetch("ENROLL_GRAPHQL_ENDPOINT", "http://localhost:3001/graphql") %>
 
-# Do not keep production secrets in the repository,
-# instead read values from the environment.
-production:
+staging:
   secret_key_base: <%= ENV["SECRET_KEY_BASE"] %>
+  api_auth_secret: <%= ENV.fetch("API_AUTH_SECRET", "no-api-secret-set") %>
+  enroll_graphql_endpoint: <%= ENV.fetch("ENROLL_GRAPHQL_ENDPOINT", "http://localhost:3001/graphql") %>
```

Comment:
@srt32 In the past I've always stored information like this in an application.yml that is gitignored. I'm curious to learn about the pros/cons/whys of this different style.  Thanks!

Reply:
Check out https://brandonhilkert.com/blog/using-rails-4-dot-1-secrets-for-configuration/ for an intro to Rails' secrets. This feature has moved around a bit in recent releases (I think it supports encryption out of the box now, which is kinda neat but we're not using that).

The pattern we try to follow here (and in the other up to date turing apps) is:
1) that all config that we want to be able to change without editing code and deploying should be read from the environment. AKA - https://12factor.net/config
2) all that env reading should happen in a centralized place upon application boot (secrets gets loaded up somewhere in the initialization process)
3) it's my personal opinion that each of these config options should have a fallback if possible. If it's a credential to a real 3rd party service we don't want to commit it but we don want devs to have as easy a time as possible to boot up the app. So, as a result, we have fallbacks in here.

Most of these points become more helpful once there is a lot more config (And things become unwieldily to manage). Check out https://github.com/turingschool/apply/blob/master/config/secrets.yml for example. We don't want each dev to have to maintain their own offline copy of that file as it changes.

This does leave the question of how should devs get real config values if they need them. An example is a stripe test token. In that case, devs can run `STRIPE_TOKEN=real token rails s` if they need only a single config or if they need a bunch of config we could use something like figaro to boot up a bunch of variables per project - https://github.com/laserlemon/figaro

phew, that turned into an essay...